File: exercicios/poo/calculadora/src/ts/shell.ts
function input() { return ""; }
export {};

// DEL!

class Calculator {
    batteryMax: number;
    battery: number;
    display: number;

    constructor(batteryMax: number) {
        this.batteryMax = batteryMax;
        this.battery = 0;
        this.display = 0;
    }

    chargeBattery(value: number): void {
        this.battery += value;
        if (this.battery > this.batteryMax) {
            this.battery = this.batteryMax;
        }
    }

    sum(a: number, b: number): void { 
        if(this.battery == 0){
            console.log("fail: bateria insuficiente");
            return;
        }
        this.battery -= 1;
        this.display = (a + b);
        //todo
    }

    division(num: number, den: number): void {
        if(this.battery == 0){
            console.log("fail: bateria insuficiente");
            return;
        }
        this.battery -= 1;
        if(den == 0){
            console.log("fail: divisao por zero");
            return;
        }
        this.display = num / den;
    }

    toString(): string {
        return `display = ${this.display.toFixed(2)}, battery = ${this.battery}`;
    }
}


// ADD!
function main() {
    // COM!
    let calc = new Calculator(0);
    // ADD!

    while (true) {
        const line = input();
        console.log("$" + line);

        const par = line.split(" ");
        const cmd = par[0];
        if (cmd == "end") {
            break;
        } 
        else if (cmd == "init") {
            // CHAME O CONSTRUTOR AQUI
            // COM!
            const batteryMax = parseInt(par[1]);
            // DEL!
            calc = new Calculator(batteryMax);
        } 
        else if (cmd == "charge") {
            // RECARREGAR A BATERIA
            // COM!
            const value = parseInt(par[1]);
            // DEL!
            calc.chargeBattery(value);
        } 
        else if (cmd == "sum") {
            // SOMAR
            // COM!
            const a = parseInt(par[1]);
            const b = parseInt(par[2]);
            // DEL!
            calc.sum(a, b);
        } 
        else if (cmd == "div") {
            // DIVIDIR
            // COM!
            const num = parseInt(par[1]);
            const den = parseInt(par[2]);
            // DEL!
            calc.division(num, den);
        } 
        else if (cmd == "show") {
            // MOSTRAR
            // DEL!
            console.log(calc.toString());
        } 
        else {
            console.log("fail: comando invalido");
        }
    }
}

main()

File: exercicios/poo/calculadora/src/java/Shell.java
import java.util.*;
import java.text.DecimalFormat;

// DEL!
class Calculator {
    public int batteryMax;
    public int battery;
    public float display;

    public Calculator(int batteryMax) {
        this.batteryMax = batteryMax;
        this.battery = 0;
        this.display = 0.0f;
    }

    public void chargeBattery(int value) {
        if (value < 0) {
            return;
        }
        this.battery += value;
        if (this.battery > this.batteryMax) {
            this.battery = this.batteryMax;
        }
    }

    public void sum(int a, int b) {
        if (this.battery == 0) {
            System.out.println("fail: bateria insuficiente");
            return;
        }
        this.battery -= 1;
        this.display = (a + b);
    }

    public void division(int num, int den) {
        if (this.battery == 0) {
            System.out.println("fail: bateria insuficiente");
            return;
        }
        this.battery -= 1;
        if (den == 0) {
            System.out.println("fail: divisao por zero");
            return;
        }
        this.display = (float) num / den;
        
    }
    public String toString() {
        DecimalFormat df = new DecimalFormat("0.00");
        return String.format("display = %s, battery = %d", df.format(this.display), this.battery);
    }
}

// ADD!

public class Shell {

    public static void main(String[] args) {
        // COM!
        Calculator calculator = new Calculator(0);
        // ADD!
        
        while (true) {
            var line = scanner.nextLine();
            System.out.println("$" + line);

            var par = line.split(" ");
            var cmd = par[0];

            if (cmd.equals("end")) {
                break;
            }
            else if (cmd.equals("init")) {
                // CHAME O CONSTRUTOR
                // COM!
                int batteryMax = Integer.parseInt(par[1]);
                // DEL!
                calculator = new Calculator(batteryMax);
            } 
            else if (cmd.equals("show")) {
                // MOSTRAR
                // DEL!
                System.out.println(calculator);
            } 
            else if (cmd.equals("charge")) {
                // RECARREGAR A BATERIA
                // COM!
                int value = Integer.parseInt(par[1]);
                // DEL!
                calculator.chargeBattery(value);
            } 
            else if (cmd.equals("sum")) {
                // SOMAR
                // COM!
                int a = Integer.parseInt(par[1]);
                int b = Integer.parseInt(par[2]);
                // DEL!
                calculator.sum(a, b);
            } 
            else if (cmd.equals("div")) {
                // DIVIDIR
                // COM!
                int num = Integer.parseInt(par[1]);
                int den = Integer.parseInt(par[2]);
                // DEL!
                calculator.division(num, den);
            } 
            else {
                System.out.println("fail: comando invalido");
            }
        }
    }

    private static Scanner scanner = new Scanner(System.in);
}
